File: src/04_grpcftp/src/grpcftpclient_main.go
/* fork from https://github.com/grpc/grpc-go/blob/master/examples/helloworld/greeter_client/main.go
 *
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

// Package main implements a client for Greeter service.
package main

import (
  "context"
  "log"
  "os"
  "time"
  "io"
  "errors"
  _"strconv"

  "google.golang.org/grpc"
  "example.com/04_grpcftp"
)

type GrpcFtpClient struct {
  grpc protos.FtpClient
}

func NewGrpcFtpClient(cc grpc.ClientConnInterface) *GrpcFtpClient {
  return &GrpcFtpClient {
    grpc: protos.NewFtpClient(cc),
  }
}

func (c *GrpcFtpClient) Hello(ctx context.Context, request string) (error, string) {
  r, err := c.grpc.Hello(ctx, &protos.HelloRequest{Message: request})
  if err != nil {
    log.Fatalf("Hello: failed0: %v", err)
    return err, ""
  }
  return nil, r.GetMessage()
}

func (c *GrpcFtpClient) List(ctx context.Context, path string, writ_cb WriteListCb) error {
  request_reader, err0 := c.grpc.List(ctx, &protos.ListRequest{File: &protos.File{Path: path, Type: protos.File_FILE, Size:0}})
  if err0 != nil {
    log.Fatalf("List: failed0: %v", err0)
    return err0
  }

  for {
    response, err := request_reader.Recv()
    if err == io.EOF {
      log.Printf("List: Done")
      return nil
    }

    if response == nil {
      err1 := errors.New("List: response is nil")
      log.Fatalf("%v", err1)
      return err1
    }

    writ_cb(response.File)
  }

  return nil
}

func (c *GrpcFtpClient) Pull(ctx context.Context, path string, write_cb WriteFileSegmentCb) error {
  request_reader, err0 := c.grpc.Pull(ctx, &protos.PullRequest{File: &protos.File{Path: path, Type: protos.File_FILE, Size:0}})
  if err0 != nil {
    log.Fatalf("Pull: failed0: %v", err0)
    return err0
  }

  read_cb := func() (*protos.FileSegment, error) {
    response, err := request_reader.Recv()
    if err == io.EOF {
      return nil, nil
    } else if err != nil {
      log.Printf("Pull: failed1: %v", err)
      return nil, err
    }
    log.Printf("Pull: %v %v", response.Segment.FileName, response.Segment.AvailableSize)
    return response.Segment, nil
  }

  if err := write_cb("pull", read_cb); err != nil {
    log.Printf("Pull: failed2: %v", err)
    return err
  }
  log.Printf("Pull: done")

  return nil
}

func (c *GrpcFtpClient) Push(ctx context.Context, path string, read_cb ReadFileSegmentCb) error {
  request_writer, err0 := c.grpc.Push(ctx)
  if err0 != nil {
    log.Fatalf("Push: failed0: %v", err0)
    return err0
  }

  request := &protos.PushRequest{
    Segment: &protos.FileSegment{},
  }

  write_cb := func(segment *protos.FileSegment) error {
    log.Printf("Push: %v %v", segment.FileName, segment.AvailableSize)
    if err := request_writer.Send(request); err != nil {
      return err
    }
    return nil
  }

  if err1 := read_cb(path, request.Segment, write_cb); err1 != nil {
    log.Printf("Push: failed1: %v", err1)
    return err1
  }

  response, err2 := request_writer.CloseAndRecv()
  if err2 != nil {
    log.Fatalf("Push: failed2: request_writer.CloseAndRecv: err:%v", err2)
    return err2
  }

  log.Printf("Push: done: response:%v", response)
  return nil
}

func main() {

  if len(os.Args) != 4 {
    log.Printf("%v localhost:50051 hello message\n", os.Args[0]);
    log.Printf("%v localhost:50051 ls from/path\n", os.Args[0]);
    log.Printf("%v localhost:50051 pull from/path\n", os.Args[0]);
    log.Printf("%v localhost:50051 push from/path\n", os.Args[0]);
    return
  }

  host := os.Args[1];
  cmd := os.Args[2];
  arg := os.Args[3];

  conn, err := grpc.Dial(host, grpc.WithInsecure(), grpc.WithBlock())
  if err != nil {
    log.Fatalf("did not connect: %v", err)
  }
  defer conn.Close()

  ctx, cancel := context.WithTimeout(context.Background(), time.Second)
  defer cancel()

  grpcFtpClient := NewGrpcFtpClient(conn)

  if cmd == "hello" {
    err, response := grpcFtpClient.Hello(ctx, arg)
    if err != nil {
      log.Printf("Hello: failed: %v", err)
      return
    }
    log.Printf("Hello: %v", response)
  } else if cmd == "ls" {
    err := grpcFtpClient.List(ctx, arg, WriteListToConsole)
    if err != nil {
      log.Printf("List: failed: %v", err)
      return
    }
  } else if cmd == "pull" {
    err := grpcFtpClient.Pull(ctx, arg, WriteFileSegmentToFile)
    if err != nil {
      log.Printf("Pull: failed: %v", err)
      return
    }
  } else if cmd == "push" {
    err := grpcFtpClient.Push(ctx, arg, ReadFileToFileSegment)
    if err != nil {
      log.Printf("Push: failed: %v", err)
      return
    }
  }
}

File: src/04_grpcftp/src/grpcftpclient_main.cc
/* fork form https://github.com/grpc/grpc/blob/master/examples/cpp/helloworld/greeter_client.cc
 *
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "grpcftp.h"

namespace GrpcFtp {

static protos::ListRequest MakeListRequest(std::string path) {
  protos::ListRequest request;
  request.mutable_file()->set_type(protos::File_Type::File_Type_DIRECTORY);
  request.mutable_file()->set_path(path);
  return request;
};

static protos::PullRequest MakePullRequest(std::string path) {
  protos::PullRequest request;
  request.mutable_file()->set_type(protos::File_Type::File_Type_FILE);
  request.mutable_file()->set_path(path);
  return request;
};

class Client {
public:
  static std::string Hello(
    std::shared_ptr<protos::Ftp::Stub> stub
      , const std::string& message) {
    protos::HelloRequest request;
    request.set_message(message);

    protos::HelloResponse response;
    grpc::ClientContext context;
    
    grpc::Status status = stub->Hello(&context, request, &response);

    // Act upon its status.
    if (status.ok()) {
      return response.message();
    } else {
      std::cout << status.error_code() << ": " << status.error_message()
                << std::endl;
      return "RPC failed";
    }
  }
 
  static grpc::Status List(std::shared_ptr<protos::Ftp::Stub> stub
      , std::string request_path
      , WriteListCb write_cb) {
    grpc::ClientContext context;
    protos::ListRequest request = MakeListRequest(request_path);
    protos::ListResponse response;
    std::unique_ptr<grpc::ClientReader<protos::ListResponse> > reponse_reader(
        stub->List(&context, request));
    while(reponse_reader->Read(&response)) {
      auto status = write_cb(response.file());
      if (!status.ok()) {
        reponse_reader->Finish();
        return status;
      }
    }
    printf("List: Done: \n");
    return reponse_reader->Finish();
  }

  static grpc::Status Pull(std::shared_ptr<protos::Ftp::Stub> stub
      , std::string &request_path
      , void *arg
      , GrpcFtp::WriteFileSegmentCb write_cb) {
    grpc::ClientContext context;
    protos::PullRequest request = MakePullRequest(request_path);
    protos::PullResponse response;

    std::unique_ptr<grpc::ClientReader<protos::PullResponse> > reponse_reader(
        stub->Pull(&context, request));

    auto read_cb = [&]() -> protos::FileSegment * {
      if (reponse_reader->Read(&response)) {
        return response.mutable_segment();
      }
      return nullptr;
    };

    auto status = write_cb(arg, read_cb);
    if (!status.ok()) {
      reponse_reader->Finish();
      return status;
    }

    printf("Pull: Done: \n");
    return reponse_reader->Finish();
  }

  static grpc::Status Push(std::shared_ptr<protos::Ftp::Stub> stub
      , std::string &request_path
      , GrpcFtp::ReadFileSegmentCb read_cb) {
    grpc::Status status;
    grpc::ClientContext context;
    protos::PushRequest request;
    protos::PushResponse response;
    protos::FileSegment *segment = request.mutable_segment();

    std::unique_ptr<grpc::ClientWriter<protos::PushRequest> > request_writer(
        stub->Push(&context, &response));

    auto write_cb = [&](protos::FileSegment *segment) -> grpc::Status {
      printf("Push:%s %lu %lu\n", segment->file_name().c_str()
          , segment->mutable_buf()->size(), segment->available_size());
      if (!request_writer->Write(request)) {
        return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "failed to write file: writer->Write err");
      }
      return grpc::Status::OK;
    };

    if (!(status = read_cb(
      segment
      , (void *) request_path.c_str()
      , write_cb
    )).ok()) {
      return status;
    }

    request_writer->WritesDone();
    status = request_writer->Finish();
    if (status.ok()) {
    }
    
    printf("Push: Done: %d\n", response.status());
    return status;
  }

};
};

int main(int argc, char** argv) {
  std::string host = "localhost:50051";
  std::string cmd;
  std::string request_path;

  if (argc != 4) {
    printf("%s localhost:50051 hello message\n", argv[0]);
    printf("%s localhost:50051 ls from/path\n", argv[0]);
    printf("%s localhost:50051 pull from/path\n", argv[0]);
    printf("%s localhost:50051 push from/path\n", argv[0]);
    return -1;
  }
  
  host = argv[1];
  cmd = argv[2];
  request_path = argv[3];
  
  std::shared_ptr<grpc::Channel> channel = grpc::CreateChannel(
      host, grpc::InsecureChannelCredentials());
  std::shared_ptr<protos::Ftp::Stub> stub(protos::Ftp::NewStub(channel));
  
  //Hello
  if (cmd.compare("hello") == 0) {
    std::string message = GrpcFtp::Client::Hello(stub, request_path);
    printf("hello: %s\n", message.c_str());
  //List directory
  } else if (cmd.compare("ls") == 0) {
    grpc::Status status = GrpcFtp::Client::List(stub, request_path, GrpcFtp::WriteListToConsole);
    if (!status.ok()) {
      std::string msg = status.error_message();
      printf("failed to FtpClient::List err:%d msg:%s\n", status.error_code(), msg.c_str());
      return -1;
    }
  //Pull a file from remote server
  } else if (cmd.compare("pull") == 0) {

    std::string save_path = "pull";    
    grpc::Status status = GrpcFtp::Client::Pull(stub
        , request_path
        , (void *)save_path.c_str()
        , GrpcFtp::WriteFileSegmentToFile);
    if (!status.ok()) {
      std::string msg = status.error_message();
      printf("failed to FtpClient::Pull err:%d msg:%s\n", status.error_code(), msg.c_str());
      return -1;
    }
  //Push a local file to server
  } else if (cmd.compare("push") == 0) {
    grpc::Status status = GrpcFtp::Client::Push(stub
        , request_path
        , GrpcFtp::ReadFileToFileSegment);
    if (!status.ok()) {
      std::string msg = status.error_message();
      printf("failed to FtpClient::Push err:%d msg:%s\n", status.error_code(), msg.c_str());
      return -1;
    }
  }

  return 0;
}

